inode2prog.cpp

```diff
@@ -73,46 +79,59 @@ int str2int (const char * ptr) {
 	return retval;
 }
 
-// Not sure, but assuming there's no more PID's than go into 64 unsigned bits..
-#define MAX_PID_LENGTH 20
+static std::string read_file (int fd) {
+	char buf[255];
+	std::string content;
 
-char * getprogname (pid_t pid) {
-	int bufsize = 80;
-	char buffer [bufsize];
+	for (int length; (length = read(fd, buf, sizeof(buf))) > 0;) {
+		if (length < 0) {
+			std::fprintf(stderr, "Error reading file: %s\n", std::strerror(errno));
+			std::exit(34);
+		}
+		content.append(buf, length);
+	}
 
-	int maxfilenamelen = 14 + MAX_PID_LENGTH + 1; 
-	char filename[maxfilenamelen];
+	return content;
```

Comment:
No, the space for `name` is allocated as part of the allocation of `newnode`. If you were to make a `prg_node` on the stack, then `name` would also live on the stack. It's just the character buffer inside of `std::string` that is dynamically allocated. The string itself is part of whatever object or stack encloses it.

Example:

    prg_node* pn = new prog_node;
    // space for name (sizeof(std::string)) is allocated as part of pn. Since name is an empty string, there will be
    // no heap allocation inside of name yet.
    // (technically there could be a heap allocation inside of name since the standard doesn't guarantee that empty
    // strings don't allocate, but every standard implementation I've seen tries very hard to avoid useless allocations.
    
    newnode->name = getprogname(pid);
    // Copy ellison pulls the content out of read_file, through getprogname, and copy-assigns it into newnode->name.

    delete pn; 
    // prog_node::~prog_node calls name::~string, releasing the buffer name holds internally

Although... thinking about this more, I'm not actually sure that gcc will ellide a copy assignment into a previously existing variable. I only know for sure that it will ellide in a case like `std::string s = some_func();`. I'm not sure about a case like `some_struct->s = some_func()`. 

It can in theory, but to do so would require it to have some `std::string` specific optimizations in it that I'm not sure it has.

All in all, I suppose if a potentially extra allocation per file name is a likely bottleneck, I need to look into this and ensure that the compile is doing what I think it is. Failing that, I can change this to be a more low-level C-esque approach. I'm just hesitant to drop to that low of level since it makes it very, very easy to make mistakes like the missing-null and memory leak that this fixes.